From 7eb0eafe29887ff7ba5488f083304e8f6cb0382e Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Wed, 3 Jul 2019 02:01:47 +0000
Subject: [PATCH] OS-6168 Attempt to first unmount "cores" before renaming the
 delegated dataset during reprovisioning Reviewed by: Jason King
 <jason.king@joyent.com> Approved by: Jason King <jason.king@joyent.com>

---
 src/vm/node_modules/VM.js | 38 +++++++++++++++++++-------------------
 1 file changed, 19 insertions(+), 19 deletions(-)

diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 7d48bbd5..28b33dfd 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -10886,25 +10886,6 @@ exports.reprovision = function (uuid, payload, options, callback)
                 }
                 cb(err);
             });
-        }, function (cb) {
-            // if we have a delegated dataset, rename zones/<uuid>/data
-            //     -> zones/<uuid>-reprovisioning-data
-            if (!vmobj.hasOwnProperty('datasets')
-                || vmobj.datasets.length === 0) {
-
-                cb();
-                return;
-            }
-            zfs(['rename', '-f', vmobj.datasets[0], vmobj.zfs_filesystem
-                + '-reprovisioning-data'], log, function (err, fds) {
-
-                if (err) {
-                    log.error({err: err, stdout: fds.stdout,
-                        stderr: fds.stderr}, 'Unable to (temporarily) rename '
-                        + vmobj.datasets[0]);
-                }
-                cb(err);
-            });
         }, function (cb) {
             // unmount <zonepath>/cores so dataset is not busy
             zfs(['umount', vmobj.zonepath + '/cores'], log,
@@ -10924,6 +10905,25 @@ exports.reprovision = function (uuid, payload, options, callback)
                 }
                 cb(err);
             });
+        }, function (cb) {
+            // if we have a delegated dataset, rename zones/<uuid>/data
+            //     -> zones/<uuid>-reprovisioning-data
+            if (!vmobj.hasOwnProperty('datasets')
+                || vmobj.datasets.length === 0) {
+
+                cb();
+                return;
+            }
+            zfs(['rename', '-f', vmobj.datasets[0], vmobj.zfs_filesystem
+                + '-reprovisioning-data'], log, function (err, fds) {
+
+                if (err) {
+                    log.error({err: err, stdout: fds.stdout,
+                        stderr: fds.stderr}, 'Unable to (temporarily) rename '
+                        + vmobj.datasets[0]);
+                }
+                cb(err);
+            });
         }, function (cb) {
             // rename <zfs_filesystem> dataset out of the way
             var cancelFn;
-- 
2.21.0

